pkg/gameserversets/gameserversets.go

```diff
@@ -75,12 +76,45 @@ func sortGameServersByLeastFullNodes(list []*agonesv1.GameServer, count map[stri
 				return false
 			}
 		}
-		// if both Nodes have the same count, one node is emptied first (packed scheduling behavior)
+
+		// Check Node total count
 		acTotal, bcTotal := ac.Allocated+ac.Ready, bc.Allocated+bc.Ready
-		if acTotal == bcTotal {
-			return a.Status.NodeName < b.Status.NodeName
+		if acTotal < bcTotal {
+			return true
 		}
-		return acTotal < bcTotal
+		if acTotal > bcTotal {
+			return false
+		}
+
+		// if the Nodes have the same count and CountsAndLists flag is enabled, sort based on CountsAndLists Priorities.
+		if runtime.FeatureEnabled(runtime.FeatureCountsAndLists) && (priorities != nil) {
+			for _, priority := range priorities {
+				res := compareGameServerCapacity(&priority, a, b)
+				switch priority.Order {
+				case agonesv1.GameServerPriorityAscending:
+					if res == -1 {
+						return true
+					}
+					if res == 1 {
+						return false
+					}
+				case agonesv1.GameServerPriorityDescending:
+					if res == -1 {
+						return false
+					}
+					if res == 1 {
+						return true
+					}
+				}
+			}
+		}
+
+		// if the gameservers are on the same node, then sort lexigraphically for a stable sort
+		if a.Status.NodeName == b.Status.NodeName {
+			return a.GetObjectMeta().GetName() < b.GetObjectMeta().GetName()
+		}
+		// if both Nodes have the same count, one node is emptied first (packed scheduling behavior)
+		return a.Status.NodeName < b.Status.NodeName
```

Comment:
Should we just rely on `GetObjectMeta().GetName()` and skip the checking of `Status.NodeName`? I.e. simplify this to:
~~~
return a.GetObjectMeta().GetName() < b.GetObjectMeta().GetName()
~~~

Reply:
No because that wouldn't give us a per-node deterministic ordering.

If two nodes have the same count, we need to always pick one over the other in a deterministic manner. This fixes a bug where concurrent processes could bounce between scaling down nodes of similar counts rather than just picking one and sticking with it.

Reply:
Gotcha, make sense.